### PR TITLE
Send span links to Cloud Trace

### LINKED ISSUE #176: Implement links in trace exporter
https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/blob/150c8dab3e9a7f336690514cfea6bf2f7b8c7826/exporter/trace/trace_proto.go#L181

### PR DESCRIPTION
**Pull request by aabmass**
Fixes #176

### CODE REVIEW

File: exporter/trace/trace_proto_test.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+				attribute.String("hello", "world"),
+			},
+		}
+		linksPb := linksProtoFromLinks([]trace.Link{link})
+
+		assert.NotNil(t, linksPb)
+		assert.EqualValues(t, linksPb.DroppedLinksCount, 0)
+		assert.Len(t, linksPb.Link, 1)
+		assert.Equal(t, spanContext.TraceID().String(), linksPb.Link[0].TraceId)
+		assert.Equal(t, spanContext.SpanID().String(), linksPb.Link[0].SpanId)
```
Comment (dashpole) [inline exporter/trace/trace_proto_test.go:52]: also assert that the type is unspecified.

File: exporter/trace/trace_proto.go [hunk 2]
```diff
@@ -178,11 +180,36 @@ func protoFromReadOnlySpan(s sdktrace.ReadOnlySpan, defaultTraceAttributes []att
...
 	}
 
-	// TODO(ymotongpoo): add implementations for Links
+	sp.Links = linksProtoFromLinks(s.Links())
 
 	return sp
 }
 
+func linksProtoFromLinks(links []trace.Link) *tracepb.Span_Links {
+	var linksPb *tracepb.Span_Links
```
Comment (dashpole) [inline exporter/trace/trace_proto.go:189]: nit: you can instantiate a return value in the function definition, and avoid the extra declaration (and you don't need to specify anything when you return):
```golang
func linksProtoFromLinks(links []trace.Link) (linksPb *tracepb.Span_Links) {
    ...
    return
}
```

**Review by dashpole — APPROVED** [review]
Looks good overall.

File: exporter/trace/trace_proto.go (hunk 2 above)
Comment (bshi) [reply exporter/trace/trace_proto.go:189]: (Sorry for the fly-by)

Please don’t use naked returns for this purpose. 

https://github.com/golang/go/wiki/CodeReviewComments#named-result-parameters

File: exporter/trace/trace_proto.go [hunk 3]
```diff
@@ -178,11 +180,36 @@ func protoFromReadOnlySpan(s sdktrace.ReadOnlySpan, defaultTraceAttributes []att
 		sp.TimeEvents.DroppedAnnotationsCount = clip32(droppedAnnotationsCount)
 	}
 
-	// TODO(ymotongpoo): add implementations for Links
+	sp.Links = linksProtoFromLinks(s.Links())
```
Comment (0xc0d) [inline exporter/trace/trace_proto.go:183]:
```suggestion
	spanLinks := s.Links()

	numLinks := len(spanLinks)
	numLinksToKeep := numLinks
	if numLinksToKeep > maxNumLinks {
		numLinksToKeep = maxNumLinks
	}

	sp.Links = linksProtoFromLinks(spanLinks[:numLinksToKeep])
	sp.Links.DroppedLinksCount = clip32(numLinks - numLinksToKeep)
```

File: exporter/trace/trace_proto.go [hunk 4]
```diff
@@ -178,11 +180,36 @@ func protoFromReadOnlySpan(s sdktrace.ReadOnlySpan, defaultTraceAttributes []att
...
+				Type:    tracepb.Span_Link_TYPE_UNSPECIFIED,
+			}
+			copyAttributes(&linkPb.Attributes, link.Attributes)
+			linksPb.Link = append(linksPb.Link, linkPb)
+		}
+		linksPb.DroppedLinksCount = clip32(numLinks - numLinksToKeep)
+	}
+
+	return linksPb
+}
```
Comment (0xc0d) [inline exporter/trace/trace_proto.go:211]: Isn't it out of `linksProtoFromLinks`'s concern to drop extra links?
```suggestion
func linksProtoFromLinks(links []trace.Link) *tracepb.Span_Links {
	if len(links) == 0 {
		return nil
	}

	linksPb := &tracepb.Span_Links{
		Link: make([]*tracepb.Span_Link, 0, len(links)),
	}
	for _, link := range links {
		linkPb := &tracepb.Span_Link{
			TraceId: link.SpanContext.TraceID().String(),
			SpanId:  link.SpanContext.SpanID().String(),
			Type:    tracepb.Span_Link_TYPE_UNSPECIFIED,
		}
		copyAttributes(&linkPb.Attributes, link.Attributes)
		linksPb.Link = append(linksPb.Link, linkPb)
	}

	return linksPb
}
```

**Review by 0xc0d — CHANGES_REQUESTED** [review]

File: exporter/trace/trace_proto.go (hunk 4 above)
Comment (aabmass) [reply exporter/trace/trace_proto.go:211]: Btw, I think ultimately we should remove all of the dropping logic from the exporter and pass [span limits](https://pkg.go.dev/go.opentelemetry.io/otel/sdk@v0.20.0/trace#WithSpanLimits) to the SDK in `NewExportPipeline()`.

File: exporter/trace/trace_proto.go (hunk 2 above)
Comment (aabmass) [reply exporter/trace/trace_proto.go:189]: I restructured this a bit

File: exporter/trace/trace_proto.go (hunk 4 above)
Comment (0xc0d) [reply exporter/trace/trace_proto.go]: Well, the function name says it is supposed to convert span links to protobuf span links, not dropping them. What you see is not what you get something.

But, yes I agree not to drop anything unless it is explicitly there.

**Review by jsuereth — APPROVED** [review]

File: exporter/trace/trace_proto.go (hunk 4 above)
Comment (aabmass) [reply exporter/trace/trace_proto.go]: @jsuereth made a good point offline that setting these limits will effect ALL exporters if the user has set up multiple pipelines. So I will keep the dropping and tweak this function name

**Pull request merged**